### PR TITLE
feat: Sync with People

### PR DESCRIPTION
**Pull request by sdsantos**

### CODE REVIEW

File: app/src/main/java/tech/relaycorp/courier/domain/server/ServerConnectionService.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    private val storedMessageDao: StoredMessageDao,
+    private val diskRepository: DiskRepository,
+    private val deleteMessage: DeleteMessage
+) : CogRPCServer.ConnectionService {
+
+    private val clientConnected = PublishChannel<Unit>()
+    fun clientConnected() = clientConnected.asFlow()
+
+    override suspend fun collectCargo(cca: CogRPC.MessageReceived): Iterable<CogRPC.MessageDelivery> {
+        clientConnected.send(Unit)
```
Comment (gnarea) [inline app/src/main/java/tech/relaycorp/courier/domain/server/ServerConnectionService.kt]: Why do we need to announce when a client connects?

Note that the server can't control the order of the methods used by the client, so even though we certainly expect this to be the first method called by the client, if we needed a reliable way to detect when a new client connects, we'd need to do it in a different way (it'll basically depend on the interface offered by the server).

File: app/src/main/java/tech/relaycorp/courier/data/disk/DiskRepository.kt [hunk 2]
```diff
@@ -34,6 +39,14 @@ class DiskRepository
         }
     }
 
+    private suspend fun writeMessageUnsafe(message: ByteArray) =
```
[the remark this answers is not in the document]
Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/courier/data/disk/DiskRepository.kt]: Going with `Unhandled`.

File: app/src/main/java/tech/relaycorp/courier/domain/PublicSync.kt [hunk 3]
```diff
@@ -18,21 +18,15 @@ class PublicSync
...
         } catch (e: CogRPCClient.Exception) {
             state.send(State.Error)
         }
     }
 
-    private suspend fun syncUnhandled() {
-        // TODO: Collect up-to-date data transferred data
-        dataTransferred.send(DataCount())
-
+    private suspend fun syncUnsafe() {
```
[the remark this answers is not in the document]
Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/courier/domain/PublicSync.kt]: Going with `Unhandled`

File: app/src/main/java/tech/relaycorp/courier/data/disk/DiskRepository.kt [hunk 4]
```diff
@@ -34,6 +39,14 @@ class DiskRepository
...
+        withContext(Dispatchers.IO) {
+            val messagesDir = getOrCreateMessagesDir()
+            val file = File.createTempFile(MESSAGE_FILE_PREFIX, "", messagesDir)
```
[the remark this answers is not in the document]
Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/courier/data/disk/DiskRepository.kt]: It actually just creates a normal file, but the `createTempFile` ensures you're getting a unique file name every time, that's why I'm using it.

File: app/src/main/java/tech/relaycorp/courier/domain/server/ServerConnectionService.kt (hunk 1 above)
Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/courier/domain/server/ServerConnectionService.kt]: I'll try to gather this information directly from the CogRPC server then.

Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/courier/domain/server/ServerConnectionService.kt]: Basically the goal was to know how many clients have synced with the Courier. But, most important is to know if there is someone still connected or not, so I'll focus on that.

File: app/src/main/java/tech/relaycorp/courier/data/disk/DiskRepository.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (gnarea) [reply app/src/main/java/tech/relaycorp/courier/data/disk/DiskRepository.kt]: Gotcha.

I assumed it worked like in other languages, where the file would be destroyed automatically when the process ends, but I just checked the docs and that isn't done automatically.

File: app/src/main/java/tech/relaycorp/courier/domain/server/ServerConnectionService.kt (hunk 1 above)
Comment (gnarea) [reply app/src/main/java/tech/relaycorp/courier/domain/server/ServerConnectionService.kt]: > Basically the goal was to know how many clients have synced with the Courier

I don't think we can do that without changes to the protocol.

> But, most important is to know if there is someone still connected or not, so I'll focus on that.

That's certainly doable. The exact way to compute that will depend on what the gRPC server gives us:

- They may keep a pool of active connections with each client, in which case it'd be a matter of counting the elements of the pool. I haven't seen this with other gRPC servers though.
- They may emit an event when a new TCP connection is established and another event when a connection ends, so we could keep a counter of those events to calculate the number of open connections. We'd just have to make sure that counter is thread-safe.

Either way, you should be able to call a method that tells you whether there's any open connection and that method could do one of those things behind the scenes.

**Comment by gnarea** [comment]
I had a look at the last commit and it looks good, but what do you think about my comments on `MessageReceived` and `MessageDeliveryAck`?

**Comment by sdsantos** [comment]
@gnarea those `MessageReceived` and `MessageDeliveryAck` are not the data structures used inside CogRPC, just external ones just to wrap data and make it clear what's the purpose of receiving them. I could just return LocalIds, but this way the classe itself says why you're getting them (to ack the delivery).

**Comment by gnarea** [comment]
Gotcha. Thanks!

**Comment by gnarea** [comment]
I think we're ready to merge this but I wasn't sure, so I'll leave it to you.

**Pull request merged**